packages/node/src/utils/kyve/kyve.ts

```diff
@@ -0,0 +1,529 @@
+        return mid;
+      }
+
+      if (height > toKey) {
+        low = mid + 1;
+      } else {
+        high = mid - 1;
+      }
+    }
+    throw new Error(`No suitable bundle found for height ${height}`);
+  }
+
+  private findBlockByHeight(
+    height: number,
+    fileCacheData: KyveBundleData[],
+  ): KyveBundleData | undefined {
+    return fileCacheData.find(
+      (bk: KyveBundleData) => bk.key === height.toString(),
+    );
+  }
+
+  private async getBundleFromCache(
+    height: number,
+  ): Promise<BundleDetails> | undefined {
+    const bundles = await this.getResolvedBundleDetails();
+    return bundles.find(
+      (b) =>
+        parseDecimal(b.from_key) <= height && height <= parseDecimal(b.to_key),
+    );
+  }
+
+  private async updateCurrentBundleAndDetails(
+    height: number,
+  ): Promise<KyveBundleData[]> {
+    let bundle = await this.getBundleFromCache(height);
+
+    if (!bundle) {
+      const bundleIds = Object.keys(this.cachedBundleDetails);
+
+      let bundleId: number;
+      if (bundleIds.length === 0) {
+        bundleId = await this.getBundleId(height);
+      } else {
+        //
+        const cachedBundles = await this.getResolvedBundleDetails();
+        const nearestBundle = cachedBundles.filter(
+          (b) => parseDecimal(b.to_key) < height,
+        );
+
+        bundleId =
+          Math.max(...nearestBundle.map((b) => parseDecimal(b.id))) + 1;
+
+        this.cachedBundleDetails[bundleId] = this.getBundleById(bundleId);
+      }
+
+      bundle = await this.cachedBundleDetails[bundleId];
+    }
+
+    return JSON.parse(await this.getBundleData(bundle));
+  }
+
+  private async pollUntilReadable(bundleFilePath: string): Promise<string> {
+    // XXXX:SCOTT This limit can be removed if the timeout problem is resolved in downloadAndProcessBundle
+    let limit = 10;
+    while (limit > 0) {
+      try {
+        return await this.readFromFile(bundleFilePath);
+      } catch (e) {
+        if (e.code === 'EACCES') {
+          await delay(POLL_TIMER);
+          limit--;
+        } else {
+          throw e;
+        }
+      }
+    }
+    throw new Error('Timeout waiting for bundle');
+  }
+
+  async downloadAndProcessBundle(bundle: BundleDetails): Promise<void> {
+    const bundleFilePath = this.getBundleFilePath(bundle.id);
+
+    const writeStream = fs.createWriteStream(bundleFilePath, {
+      flags: 'wx',
+      mode: 0o200, // Ensure that only writer has access to file
+    });
+
+    try {
+      // XXXX:SCOTT This can get stuck and not resolve, it seems a file can get stuck with permissions not reset (indexer restart)
+      // Its probably worth adding a timeout on this function
+      await Promise.race([
+        new Promise((resolve, reject) => {
+          writeStream.on('open', resolve);
+          writeStream.on('error', reject);
+        }),
+        delay(5).then(() => {
+          throw new Error('Timeout: File stream did not open');
+        }),
+      ]);
```

Comment:
We have a timeout function, but to me there is a bigger issue that this promise doesn't resolve. It should always resolve